src/library/compiler/simp_inductive.cpp

```diff
@@ -89,12 +89,17 @@ class simp_inductive_core_fn : public compiler_step_visitor {
     pair<expr, bool> visit_minor_premise(expr e, buffer<bool> const & rel_fields) {
         type_context_old::tmp_locals locals(ctx());
         for (unsigned i = 0; i < rel_fields.size(); i++) {
-            lean_assert(is_lambda(e));
-            if (rel_fields[i]) {
-                expr l = locals.push_local_from_binding(e);
-                e = instantiate(binding_body(e), l);
+            if (!is_lambda(e)) {
+                /* [note] this was added for the test `tests/lean/crash2.lean`.  */
```

Comment:
This file doesn't seem to exist.

In addition to the correct filename, could you also describe the example here?  What is the problematic term?

Looking through the file, most functions seem to assume that the minor premise is eta-expanded.  I'm surprised that just inserting a `break` here works.

Reply:
oops yes I renamed it.

The problematic term is

```
instance its_a_monad : monad option :=
{ pure := λ _, some
, bind := λ α β x f, option.rec_on x option.none f}
```
It is expecting `f` to be a lambda expression.

Reply:
I could fix it by eta expanding instead?

Reply:
No, I mean which recursor/cases_on application is the problem here?  I'd like to have the pp.all representation of the recursor term in a comment so that I know what's going on.

I believe that eta-expansion is the "proper" fix here.  But the current fix is also ok given how rare the issue is.

Reply:
The best fix would be to figure out what code further up the call stack modified the minor prem such that this routine has the right to assert that the term is a lambda, but I can't find anything that does this. I guess you would add this to `visit_cases_on`?

Reply:
Hmm, if I use `e = m_ctx.eta_expand(e);` , then test 172.lean fails to generate the bytecode:
```
failed to generate bytecode for 'its_a_monad'
unknown declaration 'its_a_monad._lambda_5'
```
If I look in the debugger, the term `e` that this method is expecting to be a lambda is:
```
lean::expr[App] = {
  m_fn = lean::expr[Constant] = {
    m_name = 'its_a_monad'.'_lambda_5',
    m_levels = lean::levels
  },
  m_arg = lean::expr[Local] = {
    m_pp_name = '_x',
    m_name = 0x0.'_fresh'.0xb.0x4d4,
    m_bi = {
      m_implicit = 0x0,
      m_strict_implicit = 0x0,
      m_inst_implicit = 0x0,
      m_rec = 0x0
    },
    m_type = lean::expr[Constant] = {
      m_name = 0x1,
      m_levels = lean::levels
    }
  }
}
```
So it looks like the compiler has already done a pass over `f`, I'm not sure where, and made an auxiliary definition for the lambda expression, which I bet will be defeq to a lambda.

Reply:
Yeah, at this point the term no longer type-checks.  So eta-expansion is probably no longer possible.

Reply:
Ok, it does feel like a hack to add in the `break; `. I'll run a build of mathlib with this change and see if it breaks anything.

Reply:
Did it break anything in mathlib?  Otherwise I'll merge it.

Reply:
Hi sorry my mathlib build kept OOMing (because my comp doesn't have enough mem) and I forgot about it. I've got a github action hopefully set up to do it now. :)